src/puya/puya-binary.ts

```diff
@@ -0,0 +1,241 @@
+    if (fs.existsSync(potentialNodeModulesDir)) {
+      return potentialNodeModulesDir
+    }
+    currentDir = path.dirname(currentDir)
+  }
+
+  // If we couldn't find it, default to the current directory
+  return process.cwd()
+}
+
+/**
+ * Finds the path to a cached Puya binary if it exists
+ * @returns The path to the cached binary or undefined if not found
+ */
+export function findCachedPuyaBinary(): string | undefined {
+  const nodeModulesDir = findNodeModulesDir()
+  const puyaStorageDir = path.join(nodeModulesDir, '.puya-ts')
+  const binaryFileName = getBinaryName()
+  const binaryPath = path.join(puyaStorageDir, binaryFileName)
+
+  return fs.existsSync(binaryPath) ? binaryPath : undefined
+}
+
+/**
+ * Deletes the cached Puya binary if it exists
+ */
+export function deleteCachedPuyaBinary(): void {
+  const cachedPath = findCachedPuyaBinary()
+  if (cachedPath && fs.existsSync(cachedPath)) {
+    try {
+      logger.info(undefined, `Removing outdated Puya binary at ${cachedPath}`)
+      fs.unlinkSync(cachedPath)
+    } catch (error) {
+      logger.warn(undefined, `Failed to delete outdated Puya binary: ${error instanceof Error ? error.message : String(error)}`)
+      throw error
+    }
+  }
+}
+
+/**
+ * Gets the platform-specific details for downloads
+ * @returns Object containing OS and architecture information
+ */
+function getPlatformDetails(): { os: string; arch: string } {
+  // Map Node.js platform to OS name used in filenames
+  let os: string
+  switch (process.platform) {
+    case 'win32':
+      os = 'windows'
+      break
+    case 'darwin':
+      os = 'macos'
+      break
+    case 'linux':
+      os = 'linux'
+      break
+    default:
+      throw new Error(`Unsupported platform: ${process.platform}`)
+  }
+
+  // Map Node.js architecture to architecture name used in filenames
+  let arch: string
+  switch (process.arch) {
+    case 'x64':
+      arch = 'x64'
+      break
+    case 'arm64':
+      arch = 'arm64'
+      break
+    default:
+      throw new Error(`Unsupported architecture: ${process.arch}`)
+  }
+
+  return { os, arch }
+}
+
+/**
+ * Downloads the Puya binary for a specific release version
+ * @param version The release version to download (e.g., "v1.0.0")
+ * @returns Promise that resolves to the path of the extracted binary
+ */
+export async function downloadPuyaBinary(version: string): Promise<string> {
+  // Get platform-specific details
+  const { os, arch } = getPlatformDetails()
+
+  // Build platform-specific filenames
+  const platformId = `${os}_${arch}`
+  const archiveFileName = `puya-${platformId}.tar.gz`
+  const checksumFileName = `puya-${platformId}-checksum.txt`
+  const binaryFileName = getBinaryName()
+
+  // Find node_modules directory and set up storage paths
+  const nodeModulesDir = findNodeModulesDir()
+  const puyaStorageDir = path.join(nodeModulesDir, '.puya-ts')
+  const tempDir = path.join(nodeModulesDir, '.puya-ts-temp')
```

Comment:
We actually have a disposable type for managing temp dirs.

```suggestion
  const tempDir = path.join(nodeModulesDir, '.puya-ts-temp')
  using tempDir = generateTempDir()
```

If you use this you can get rid of some of the superfluous try/catch blocks. 

there is also a temp file if you don't have strong opinions on the name of the file that is created. (and the file is only going to exist temporarily)

Reply:
this is cool, I updated the code.